### DISCUSSION TITLE
아이템 12. toString을 항상 재정의하라

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Effective-Java/Effective-Java#32] [3. 과제]
## `toString()` 메서드 특징

- `Object` 내에 속한 메서드로 **클래스_이름@16진수로_표시한_해시코드**
- `toString` 메서드는 **간결하면서 사람이 읽기 쉬운 형태의 유익한 정보** 를 반환해야 한다.
- 클라이언트가 직접 호출하지 않아도, 다른 어딘가에서 쓰일 수 있음
    e.g. `println`, `printf`, 문자열 연결 연산자(`+`), `assert` 구문, ...

## 좋은 `toString()` 작성 요령

- 객체가 가진 모든 정보를 보여주는 것이 좋다.
- 값 클래스라면 `toString()` 포맷을 문서에 명시하는 것이 좋다.
- 해당 포맷으로 객체를 생성할 수 있는 정적 팩터리나 생성자를 제공하는 것이 좋다.
- `toString()` 메서드가 반환한 값에 포함된 정보를 얻어올 수 있는 API 를 제공하는 것이 좋다.
- 우리가 원하는 포맷이 필요할 경우에 *AutoValue*, *Lombok*, *IDE* 사용이 능사는 아니다.

```java
import java.util.Objects;

public class PhoneNumber {

    private final int areaCode;
    private final int prefix;
    private final int lineNumber;

    public PhoneNumber(int areaCode, int prefix, int lineNumber) {
        this.areaCode = areaCode;
        this.prefix = prefix;
        this.lineNumber = lineNumber;
    }

    public static PhoneNumber of(String phoneNumberAsString) {
        String[] numberTokens = phoneNumberAsString.split("-");

        return new PhoneNumber(Integer.parseInt(numberTokens[0]),
                               Integer.parseInt(numberTokens[1]),
                               Integer.parseInt(numberTokens[2]));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PhoneNumber that = (PhoneNumber) o;
        return areaCode == that.areaCode && prefix == that.prefix && lineNumber == that.lineNumber;
    }

    @Override
    public int hashCode() {
        return Objects.hash(areaCode, prefix, lineNumber);
    }

}
```

```java
public class PhoneNumberShell {

    public static void main(String[] args) {
        PhoneNumber cherryPhoneNumber = new PhoneNumber(010, 8765, 4321);
        System.out.println("Cherry's phone number: " + cherryPhoneNumber);

        PhoneNumber target = PhoneNumber.of("010-8765-4321");
        System.out.println(target);
        System.out.println(cherryPhoneNumber.equals(target));
        System.out.println(cherryPhoneNumber.hashCode() == target.hashCode());
    }

}
```

## AutoValue

- AutoValue 는 Google 에서 개발한 코드 자동 생성기로 자바 코드를 줄이는 데 기여한다.
- *Reflection* 방식을 이용한 런타임 방식이 아닌 *apt* 를 이용하여 컴파일 타임에 코드를 생성한다.
- MVN Repository 에서 AutoValue jar 파일을 다운로드 후 사용한다.
- [`@AutoValue`](https://mvnrepository.com/artifact/com.google.auto.value/auto-value) 애노테이션을 사용하여 코드를 작성할 수 있다.

## Lombok `@ToString`

- 자바에서 반복되는 소스 코드를 컴파일 과정에서 생성해주는 라이브러리다.
- [`@ToString`](https://projectlombok.org/api/lombok/ToString) 애노테이션을 통해 코드를 작성할 수 있다.

## IDE

IntelliJ 의 경우 IDE 의 기능을 이용하여 `toString()` 메서드를 오버라이딩(*overriding*)할 수 있다.

```java
public class PhoneNumber {

    // (...)

    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNumber=" + lineNumber +
            '}';
    }

}
```

**Comment by Irisation23** [comment]
정리 해 주신 글 잘 읽었습니다. 👍
`@AutoValue` 같은 경우는 조금 생소해서 찾아보니 예제를 통해 알아 볼 수있는 자료가 있어서 공유하고자 올리게 되었습니다.
> - https://www.baeldung.com/introduction-to-autovalue
> 해당 자료는 `@AutoValue`의 역할에 대해 알아 볼 수 있습니다.

조금 더 찾다보니 아래의 자료를 찾게 되었는데 Lombok vs AutoValue vs Immutables 가 같이 비교 된다는 자료를 보게 되었습니다.
각각의 라이브러리가 차이가 있고 장단점이 있다는 판단은 드는데 어느 시점에 어떤 라이브러리를 이용 하는게 가장 좋을지에 대해 알고 싶습니다. 
> - https://dev.to/cchacin/immutables-autovalue-lombok-which-one-2j33
> 해당 자료는 `Lombok vs AutoValue vs Immutables` 의 비교에 대해 알아 볼 수 있습니다.

**Reply by coalong** [reply]
`@AutoValue` 에 대해 좋은 자료를 공유해주셔서 이해하는데 도움이 됐어요! 감사합니다!!

**Reply by chikeem90** [reply]
> 각각의 라이브러리가 차이가 있고 장단점이 있다는 판단은 드는데 어느 시점에 어떤 라이브러리를 이용 하는게 가장 좋을지에 대해 알고 싶습니다.

@bunsung92 님
주신 링크에는 Immutables가 가장 좋다 라고 써있는 거 같긴한데,, ㅎㅎ
AutoValue, Immutables <> Lombok 이 역할이 나뉘는 거 같아요
AutoValue, Immutables은 value class 만들 때 사용하기 적합한 거 같고,,(Builder, equals, hashCode 정도 자동생성) Lombok은 자바 코드 자동 생성 라이브러리 정도로 보면 될 거 같네요.
그래서 사실 제공하는 기능들은 Lombok이 훨씬 많고,, 그래서 더 많이 사용되는 거 같아요.
그리고 AutoValue는 적용하려면 abstract class로 정의해야만 하는데 lombok은 그런 제한도 없어서,, 사용하기 더 편한 거 같아요.

https://mvnrepository.com/open-source/code-generators
maven 레포지토리에서도 보면 lombok이 가장 많이 사용하고 있고 인기가 제일 많네요 ㅎㅎ

**Comment by JoisFe** [comment]
toString 또한 제가 필요한 상황에만 intelliJ의 IDE 기능을 통해 생각 없이 만드는 좋지 못한 습관을 가지고 있었습니다.
하지만 이번 글을 통해 toString에 대해 좀 더 알 수 있었고 특히 toString의 작성요령에 대해 알게 되어 좋았습니다.
그런데 작성요령 중에
### 해당 포맷으로 객체를 생성할 수 있는 정적 팩터리나 생성자를 제공하는 것이 좋다.
이 부분이 toString 작성요령에 왜 굳이 포함되어있는지 잘 이해가 가지 않습니다. 해당 이유를 아신다면 공유해주시면 감사하겠습니다.
자료 정리 하신다고 고생 많으셨습니다. ^^

**Reply by JoisFe** [reply]
해당 포맷에 대한 클래스 타입을 따로 만들고 toString 반환시 해당 포멧에 대한 클래스의 객체를 반환하기 위해
해당 클래스에 생성자, 정적 팩터리로 이해했는데 맞을까요?

**Comment by chikeem90** [comment]
> Reflection 방식을 이용한 런타임 방식이 아닌 apt 를 이용하여 컴파일 타임에 코드를 생성한다.

apt가,, 뭘까요?
